pkg/operator/sync.go

```diff
@@ -0,0 +1,303 @@
+	errs = append(errs, optr.syncMachineConfigPools(*config))
+	errs = append(errs, optr.syncMachineConfigController(*config))
+	errs = append(errs, optr.syncMachineConfigDaemon(*config))
+
+	agg := utilerrors.NewAggregate(errs)
+	if agg != nil {
+		return fmt.Errorf("error syncing: %v", agg.Error())
+	}
+	return nil
+}
+
+func (optr *Operator) syncCustomResourceDefinitions(config renderConfig) error {
+	mcBytes, err := renderAsset(config, "manifests/machineconfig.crd.yaml")
+	if err != nil {
+		return err
+	}
+	ccBytes, err := renderAsset(config, "manifests/controllerconfig.crd.yaml")
+	if err != nil {
+		return err
+	}
+	mcpBytes, err := renderAsset(config, "manifests/machineconfigpool.crd.yaml")
+	if err != nil {
+		return err
+	}
+
+	updated := false
+	mc := resourceread.ReadCustomResourceDefinitionV1Beta1OrDie(mcBytes)
+	_, updated, err = resourceapply.ApplyCustomResourceDefinition(optr.apiExtClient.ApiextensionsV1beta1(), mc)
+	if err != nil {
+		return err
+	}
+	if updated {
+		if err := optr.waitForCustomResourceDefinition(mc); err != nil {
+			return err
+		}
+	}
+
+	cc := resourceread.ReadCustomResourceDefinitionV1Beta1OrDie(ccBytes)
+	_, updated, err = resourceapply.ApplyCustomResourceDefinition(optr.apiExtClient.ApiextensionsV1beta1(), cc)
+	if err != nil {
+		return err
+	}
+	if updated {
+		if err := optr.waitForCustomResourceDefinition(cc); err != nil {
+			return err
+		}
+	}
+
+	mcp := resourceread.ReadCustomResourceDefinitionV1Beta1OrDie(mcpBytes)
+	_, updated, err = resourceapply.ApplyCustomResourceDefinition(optr.apiExtClient.ApiextensionsV1beta1(), mcp)
+	if err != nil {
+		return err
+	}
+	if updated {
+		if err := optr.waitForCustomResourceDefinition(mcp); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (optr *Operator) syncMachineConfigPools(config renderConfig) error {
+	masterBytes, err := renderAsset(config, "manifests/master.machineconfigpool.yaml")
+	if err != nil {
+		return err
+	}
+	workerBytes, err := renderAsset(config, "manifests/worker.machineconfigpool.yaml")
+	if err != nil {
+		return err
+	}
+	etcdBytes, err := renderAsset(config, "manifests/etcd.machineconfigpool.yaml")
+	if err != nil {
+		return err
+	}
+
+	master := resourceread.ReadMachineConfigPoolV1OrDie(masterBytes)
+	_, _, err = resourceapply.ApplyMachineConfigPool(optr.client.MachineconfigurationV1(), master)
+	if err != nil {
+		return err
+	}
+	worker := resourceread.ReadMachineConfigPoolV1OrDie(workerBytes)
+	_, _, err = resourceapply.ApplyMachineConfigPool(optr.client.MachineconfigurationV1(), worker)
+	if err != nil {
+		return err
+	}
+	etcd := resourceread.ReadMachineConfigPoolV1OrDie(etcdBytes)
+	_, _, err = resourceapply.ApplyMachineConfigPool(optr.client.MachineconfigurationV1(), etcd)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (optr *Operator) syncMachineConfigController(config renderConfig) error {
+	ccBytes, err := renderAsset(config, "manifests/machineconfigcontroller/controllerconfig.yaml")
```

Comment:
same here, we can create a map from the asset name to their `render` and `read` functions, and potentially wait for the resources. Should be pretty generic to reduce the code redundancy .

Reply:
read* and apply* expect specific types. 
Will end up doing bunch of type assertions.